File: ios/Classes/ClickstreamFlutterPlugin.swift
//
// Copyright Amazon.com Inc. or its affiliates.
// All Rights Reserved.
//
// SPDX-License-Identifier: Apache-2.0
//

import Amplify
import Flutter
import UIKit

public class ClickstreamFlutterPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "clickstream_flutter", binaryMessenger: registrar.messenger())
        let instance = ClickstreamFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "init":
            result(initSDK(call.arguments as! [String: Any]))
        case "record":
            recordEvent(call.arguments as! [String: Any])
        case "setUserId":
            setUserId(call.arguments as! [String: Any])
        case "setUserAttributes":
            setUserAttributes(call.arguments as! [String: Any])
        case "addGlobalAttributes":
            addGlobalAttributes(call.arguments as! [String: Any])
        case "deleteGlobalAttributes":
            deleteGlobalAttributes(call.arguments as! [String: Any])
        case "updateConfigure":
            updateConfigure(call.arguments as! [String: Any])
        case "flushEvents":
            ClickstreamAnalytics.flushEvents()
        case "disable":
            ClickstreamAnalytics.disable()
        case "enable":
            ClickstreamAnalytics.enable()
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    func initSDK(_ arguments: [String: Any]) -> Bool {
        do {
            let configuration = ClickstreamConfiguration()
                .withAppId(arguments["appId"] as! String)
                .withEndpoint(arguments["endpoint"] as! String)
                .withLogEvents(arguments["isLogEvents"] as! Bool)
                .withTrackScreenViewEvents(arguments["isTrackScreenViewEvents"] as! Bool)
                .withTrackUserEngagementEvents(arguments["isTrackUserEngagementEvents"] as! Bool)
                .withTrackAppExceptionEvents(arguments["isTrackAppExceptionEvents"] as! Bool)
                .withSendEventInterval(arguments["sendEventsInterval"] as! Int)
                .withSessionTimeoutDuration(arguments["sessionTimeoutDuration"] as! Int64)
                .withCompressEvents(arguments["isCompressEvents"] as! Bool)
                .withAuthCookie(arguments["authCookie"] as! String)
            if let attributes = arguments["globalAttributes"] as? [String: Any] {
                if attributes.count > 0 {
                    let globalAttributes = getClickstreamAttributes(attributes)
                    _ = configuration.withInitialGlobalAttributes(globalAttributes)
                }
            }
            try ClickstreamAnalytics.initSDK(configuration)
            return true
        } catch {
            log.error("Fail to initialize ClickstreamAnalytics: \(error)")
            return false
        }
    }

    func recordEvent(_ arguments: [String: Any]) {
        let eventName = arguments["eventName"] as! String
        let attributes = arguments["attributes"] as! [String: Any]
        let items = arguments["items"] as! [[String: Any]]
        if attributes.count > 0 {
            if items.count > 0 {
                var clickstreamItems: [ClickstreamAttribute] = []
                for itemObject in items {
                    clickstreamItems.append(getClickstreamAttributes(itemObject))
                }
                ClickstreamAnalytics.recordEvent(eventName, getClickstreamAttributes(attributes), clickstreamItems)
            } else {
                ClickstreamAnalytics.recordEvent(eventName, getClickstreamAttributes(attributes))
            }
        } else {
            ClickstreamAnalytics.recordEvent(eventName)
        }
    }

    func setUserId(_ arguments: [String: Any]) {
        if arguments["userId"] is NSNull {
            ClickstreamAnalytics.setUserId(nil)
        } else {
            ClickstreamAnalytics.setUserId(arguments["userId"] as? String)
        }
    }

    func setUserAttributes(_ arguments: [String: Any]) {
        ClickstreamAnalytics.addUserAttributes(getClickstreamAttributes(arguments))
    }

    func addGlobalAttributes(_ arguments: [String: Any]) {
        ClickstreamAnalytics.addGlobalAttributes(getClickstreamAttributes(arguments))
    }

    func deleteGlobalAttributes(_ arguments: [String: Any]) {
        let attributes = arguments["attributes"] as! [String]
        for attribute in attributes {
            ClickstreamAnalytics.deleteGlobalAttributes(attribute)
        }
    }

    func updateConfigure(_ arguments: [String: Any]) {
        do {
            let configure = try ClickstreamAnalytics.getClickstreamConfiguration()
            if let appId = arguments["appId"] as? String {
                configure.appId = appId
            }
            if let endpoint = arguments["endpoint"] as? String {
                configure.endpoint = endpoint
            }
            if let isLogEvents = arguments["isLogEvents"] as? Bool {
                configure.isLogEvents = isLogEvents
            }
            if let isTrackScreenViewEvents = arguments["isTrackScreenViewEvents"] as? Bool {
                configure.isTrackScreenViewEvents = isTrackScreenViewEvents
            }
            if let isTrackUserEngagementEvents = arguments["isTrackUserEngagementEvents"] as? Bool {
                configure.isTrackUserEngagementEvents = isTrackUserEngagementEvents
            }
            if let isTrackAppExceptionEvents = arguments["isTrackAppExceptionEvents"] as? Bool {
                configure.isTrackAppExceptionEvents = isTrackAppExceptionEvents
            }
            if let sessionTimeoutDuration = arguments["sessionTimeoutDuration"] as? Int64 {
                configure.sessionTimeoutDuration = sessionTimeoutDuration
            }
            if let isCompressEvents = arguments["isCompressEvents"] as? Bool {
                configure.isCompressEvents = isCompressEvents
            }
            if let authCookie = arguments["authCookie"] as? String {
                configure.authCookie = authCookie
            }
        } catch {
            log.error("Failed to config ClickstreamAnalytics: \(error)")
        }
    }

    func getClickstreamAttributes(_ attrs: [String: Any]) -> ClickstreamAttribute {
        var attributes: ClickstreamAttribute = [:]
        for (key, value) in attrs {
            if value is String {
                attributes[key] = value as! String
            } else if value is NSNumber {
                let value = value as! NSNumber
                let objCType = String(cString: value.objCType)
                if objCType == "c" {
                    attributes[key] = value.boolValue
                } else if objCType == "d" {
                    attributes[key] = value.doubleValue
                } else if objCType == "i" {
                    attributes[key] = value.intValue
                } else if objCType == "q" {
                    attributes[key] = value.int64Value
                }
            }
        }
        return attributes
    }
}

extension ClickstreamFlutterPlugin: DefaultLogger {}

File: android/src/main/kotlin/software/aws/solution/clickstream_analytics/ClickstreamFlutterPlugin.kt
/*
 * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
package software.aws.solution.clickstream_analytics

import android.app.Activity
import com.amazonaws.logging.Log
import com.amazonaws.logging.LogFactory
import com.amplifyframework.core.Amplify
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import software.aws.solution.clickstream.ClickstreamAnalytics
import software.aws.solution.clickstream.ClickstreamAttribute
import software.aws.solution.clickstream.ClickstreamConfiguration
import software.aws.solution.clickstream.ClickstreamEvent
import software.aws.solution.clickstream.ClickstreamItem
import software.aws.solution.clickstream.ClickstreamUserAttribute
import software.aws.solution.clickstream.client.util.ThreadUtil
import java.util.Objects
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors


/** ClickstreamFlutterPlugin */
class ClickstreamFlutterPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity

    private val cachedThreadPool: ExecutorService by lazy { Executors.newCachedThreadPool() }

    private val log: Log = LogFactory.getLog(
        ClickstreamFlutterPlugin::class.java
    )
    private lateinit var channel: MethodChannel

    private var mActivity: Activity? = null

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "clickstream_flutter")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        val arguments = call.arguments() as HashMap<String, Any>?
        when (call.method) {
            "init" -> result.success(initSDK(arguments!!))
            "record" -> recordEvent(arguments)
            "setUserId" -> setUserId(arguments)
            "setUserAttributes" -> setUserAttributes(arguments)
            "addGlobalAttributes" -> addGlobalAttributes(arguments)
            "deleteGlobalAttributes" -> deleteGlobalAttributes(arguments)
            "updateConfigure" -> updateConfigure(arguments)
            "flushEvents" -> ClickstreamAnalytics.flushEvents()
            "disable" -> ClickstreamAnalytics.disable()
            "enable" -> ClickstreamAnalytics.enable()
            else -> result.notImplemented()
        }
    }

    private fun initSDK(arguments: HashMap<String, Any>): Boolean {
        if (getIsInitialized()) return false
        if (mActivity != null) {
            val context = mActivity!!.applicationContext
            if (ThreadUtil.notInMainThread()) {
                log.error("Clickstream SDK initialization failed, please initialize in the main thread")
                return false
            }
            val sessionTimeoutDuration = arguments["sessionTimeoutDuration"]
                .let { (it as? Int)?.toLong() ?: (it as Long) }
            val sendEventsInterval = arguments["sendEventsInterval"]
                .let { (it as? Int)?.toLong() ?: (it as Long) }
            val configuration = ClickstreamConfiguration()
                .withAppId(arguments["appId"] as String)
                .withEndpoint(arguments["endpoint"] as String)
                .withLogEvents(arguments["isLogEvents"] as Boolean)
                .withTrackScreenViewEvents(arguments["isTrackScreenViewEvents"] as Boolean)
                .withTrackUserEngagementEvents(arguments["isTrackUserEngagementEvents"] as Boolean)
                .withTrackAppExceptionEvents(arguments["isTrackAppExceptionEvents"] as Boolean)
                .withSendEventsInterval(sendEventsInterval)
                .withSessionTimeoutDuration(sessionTimeoutDuration)
                .withCompressEvents(arguments["isCompressEvents"] as Boolean)
                .withAuthCookie(arguments["authCookie"] as String)

            (arguments["globalAttributes"] as? HashMap<*, *>)?.takeIf { it.isNotEmpty() }
                ?.let { attributes ->
                    val globalAttributes = ClickstreamAttribute.builder()
                    for ((key, value) in attributes) {
                        when (value) {
                            is String -> globalAttributes.add(key.toString(), value)
                            is Double -> globalAttributes.add(key.toString(), value)
                            is Boolean -> globalAttributes.add(key.toString(), value)
                            is Int -> globalAttributes.add(key.toString(), value)
                            is Long -> globalAttributes.add(key.toString(), value)
                        }
                    }
                    configuration.withInitialGlobalAttributes(globalAttributes.build())
                }
            return try {
                ClickstreamAnalytics.init(context, configuration)
                true
            } catch (exception: Exception) {
                log.error("Clickstream SDK initialization failed with error: " + exception.message)
                false
            }
        } else {
            return false
        }
    }

    private fun recordEvent(arguments: HashMap<String, Any>?) {
        cachedThreadPool.execute {
            arguments?.let {
                val eventName = it["eventName"] as String
                val attributes = it["attributes"] as HashMap<*, *>
                val items = it["items"] as ArrayList<*>
                val eventBuilder = ClickstreamEvent.builder().name(eventName)
                for ((key, value) in attributes) {
                    when (value) {
                        is String -> eventBuilder.add(key.toString(), value)
                        is Double -> eventBuilder.add(key.toString(), value)
                        is Boolean -> eventBuilder.add(key.toString(), value)
                        is Int -> eventBuilder.add(key.toString(), value)
                        is Long -> eventBuilder.add(key.toString(), value)
                    }
                }
                if (items.size > 0) {
                    val clickstreamItems = arrayOfNulls<ClickstreamItem>(items.size)
                    for (index in 0 until items.size) {
                        val builder = ClickstreamItem.builder()
                        for ((key, value) in (items[index] as HashMap<*, *>)) {
                            when (value) {
                                is String -> builder.add(key.toString(), value)
                                is Double -> builder.add(key.toString(), value)
                                is Boolean -> builder.add(key.toString(), value)
                                is Int -> builder.add(key.toString(), value)
                                is Long -> builder.add(key.toString(), value)
                            }
                        }
                        clickstreamItems[index] = builder.build()
                    }
                    eventBuilder.setItems(clickstreamItems)
                }
                ClickstreamAnalytics.recordEvent(eventBuilder.build())
            }
        }
    }


    private fun setUserId(arguments: java.util.HashMap<String, Any>?) {
        arguments?.let {
            val userId = arguments["userId"]
            if (userId == null) {
                ClickstreamAnalytics.setUserId(null)
            } else {
                ClickstreamAnalytics.setUserId(userId.toString())
            }
        }
    }

    private fun setUserAttributes(arguments: java.util.HashMap<String, Any>?) {
        arguments?.let {
            val builder = ClickstreamUserAttribute.Builder()
            for ((key, value) in arguments) {
                when (value) {
                    is String -> builder.add(key, value)
                    is Double -> builder.add(key, value)
                    is Boolean -> builder.add(key, value)
                    is Int -> builder.add(key, value)
                    is Long -> builder.add(key, value)
                }
            }
            ClickstreamAnalytics.addUserAttributes(builder.build())
        }
    }

    private fun addGlobalAttributes(arguments: java.util.HashMap<String, Any>?) {
        arguments?.let {
            val builder = ClickstreamAttribute.Builder()
            for ((key, value) in arguments) {
                when (value) {
                    is String -> builder.add(key, value)
                    is Double -> builder.add(key, value)
                    is Boolean -> builder.add(key, value)
                    is Int -> builder.add(key, value)
                    is Long -> builder.add(key, value)
                }
            }
            ClickstreamAnalytics.addGlobalAttributes(builder.build())
        }
    }

    private fun deleteGlobalAttributes(arguments: java.util.HashMap<String, Any>?) {
        arguments?.let {
            @Suppress("UNCHECKED_CAST")
            val attributes = arguments["attributes"] as ArrayList<String>
            ClickstreamAnalytics.deleteGlobalAttributes(*attributes.toTypedArray())
        }
    }

    private fun updateConfigure(arguments: java.util.HashMap<String, Any>?) {
        arguments?.let {
            val configure = ClickstreamAnalytics.getClickStreamConfiguration()
            arguments["appId"]?.let {
                configure.withAppId(it as String)
            }
            arguments["endpoint"]?.let {
                configure.withEndpoint(it as String)
            }
            arguments["isLogEvents"]?.let {
                configure.withLogEvents(it as Boolean)
            }
            arguments["isTrackScreenViewEvents"]?.let {
                configure.withTrackScreenViewEvents(it as Boolean)
            }
            arguments["isTrackUserEngagementEvents"]?.let {
                configure.withTrackUserEngagementEvents(it as Boolean)
            }
            arguments["isTrackAppExceptionEvents"]?.let {
                configure.withTrackAppExceptionEvents(it as Boolean)
            }
            arguments["sessionTimeoutDuration"]?.let {
                val sessionTimeoutDuration = arguments["sessionTimeoutDuration"]
                    .let { (it as? Int)?.toLong() ?: (it as Long) }
                configure.withSessionTimeoutDuration(sessionTimeoutDuration)
            }
            arguments["isCompressEvents"]?.let {
                configure.withCompressEvents(it as Boolean)
            }
            arguments["authCookie"]?.let {
                configure.withAuthCookie(it as String)
            }
        }
    }

    private fun getIsInitialized(): Boolean {
        return invokeSuperMethod(Amplify.Analytics, "isConfigured") as Boolean
    }

    @Throws(Exception::class)
    fun invokeSuperMethod(`object`: Any, methodName: String): Any? {
        val method =
            Objects.requireNonNull(`object`.javaClass.superclass).getDeclaredMethod(methodName)
        method.isAccessible = true
        return method.invoke(`object`)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        mActivity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        mActivity = binding.activity
    }

    override fun onDetachedFromActivity() {
        mActivity = null
    }
}
